pinot-core/src/main/java/org/apache/pinot/core/routing/TablePartitionInfo.java

```diff
@@ -29,15 +32,22 @@ public class TablePartitionInfo {
   private final int _numPartitions;
   private final PartitionInfo[] _partitionInfoMap;
   private final List<String> _segmentsWithInvalidPartition;
+  /**
+   * New segments may not have all replicas available soon after creation, and hence they can potentially reduce
+   * the fullyReplicatedServers in PartitionInfo. To prevent that, we store such segments separately in this map.
+   */
+  private final Map<Integer, List<String>> _excludedNewSegments;
 
   public TablePartitionInfo(String tableNameWithType, String partitionColumn, String partitionFunctionName,
-      int numPartitions, PartitionInfo[] partitionInfoMap, List<String> segmentsWithInvalidPartition) {
+      int numPartitions, PartitionInfo[] partitionInfoMap, List<String> segmentsWithInvalidPartition,
+      Map<Integer, List<String>> excludedNewSegments) {
```

Comment:
nit: mark the existing constructor with @VisibleForTesting and create a new constructor to pass in the excludedNewSegments. The modification of existing unit test of TablePartitionInfo don't need to be modified in this way.

Reply:
imo having multiple ctors is more error prone in the long run. It's easy to call ctors with lower number of args and developers then don't think about everything that's required to create an accurate version of an object.